docs/contributor/arch/013-leader-receiver-creator-as-extension.md

```diff
@@ -0,0 +1,56 @@
+# 13. Leader Receiver Creator as Extension
+
+Date: 2024-11-08
+
+## Status
+Proposed
+
+## Context
+
+Telemetry OTeL collector collects metrics exposed by the Kubernetes API server using [k8sclusterreceiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/k8sclusterreceiver), 
+and custom resource metrics are collected using [kymastatsreceiver](https://github.com/kyma-project/opentelemetry-collector-components/tree/main/receiver/kymastatsreceiver). 
+To run these receivers in high availability mode and prevent sending duplicate metrics, we've implemented [singletonreceivercreator](https://github.com/kyma-project/opentelemetry-collector-components/tree/main/receiver/singletonreceivercreator) based on [leader receiver creator](./012-leader-receiver-creator.md).
+
+## Decision
+However, after feedback from community and careful deliberation, the `singleonreceivercreator` would be better suited to be used as an extension.
+[Extensions](https://github.com/open-telemetry/opentelemetry-collector/blob/main/extension/README.md?plain=1) provide additional functionality to the collector, but do not need direct access to telemetry data and are not part of the pipelines.
+
+Implementing `singleonreceivercreator` as an extension brings the following advantages:
+- It is signal-agnostic.
+- It can be used with any receiver.
+- It brings simpler and clear configuration.
+- It's future-proof.
+
+It has following disadvantages:
+- The receiver must be modified to support the leader election extension.
```

Comment:
- Which "receiver"? k8sclusterreceiver or kymastatsreceiver, or both?
- Which "leader election extension"? The one we're about to create, meaning Singleton Receiver Creator as extension (instead of as part of the pipeline)? Will this extension be called Singleton Receiver Creator or get another name (like "leader election extension")? I'm slightly confused by the terminology here.

Reply:
any receiver that wants to use this extension, has to be modified to have support for the extension. Currently we have 2 receivers that would require this support. K8sClusterReceiver and Kyma Stats Receiver. Changes to Kyma stats receiver is easy as its maintained by us. Changes to K8sClusterReceiver is slightly tricky as its upstream receiver and changes there would take quite sometime to be incorprated and could be used in Kyma again.

Reply:
Then plural would be good

```suggestion
- The receivers must be modified to support the leader election extension.
```

Furthermore:
Which "leader election extension"? The one we're about to create, meaning Singleton Receiver Creator as extension (instead of as part of the pipeline)? Will this extension be called Singleton Receiver Creator or get another name (like "leader election extension")? I'm slightly confused by the terminology here.

Reply:
The name of the extension we thought is leaderelector. Singleton receiver was a receiver which will be deprecated and removed in favour of leaderelector. Leader elector is an extension which implements leader election mechanism (something native to kubernetes).